src/lib/operators/join_nested_loop.cpp

```diff
@@ -0,0 +1,239 @@
+#include "join_nested_loop.hpp"
+
+#include <map>
+#include <memory>
+#include <numeric>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "resolve_type.hpp"
+#include "storage/iterables/create_iterable_from_column.hpp"
+#include "type_comparison.hpp"
+#include "utils/assert.hpp"
+#include "utils/performance_warning.hpp"
+
+namespace opossum {
+
+/*
+ * This is a Nested Loop Join implementation completely based on iterables.
+ * It supports all current join and scan types, as well as NULL values.
+ * Because this is a Nested Loop Join, the performance is going to be far inferior to JoinHash and JoinSortMerge,
+ * so only use this for testing or benchmarking purposes.
+ */
+
+JoinNestedLoop::JoinNestedLoop(const std::shared_ptr<const AbstractOperator> left,
+                               const std::shared_ptr<const AbstractOperator> right, const JoinMode mode,
+                               const std::pair<ColumnID, ColumnID>& column_ids, const ScanType scan_type)
+    : AbstractJoinOperator(left, right, mode, column_ids, scan_type) {}
+
+const std::string JoinNestedLoop::name() const { return "JoinNestedLoop"; }
+
+std::shared_ptr<AbstractOperator> JoinNestedLoop::recreate(const std::vector<AllParameterVariant>& args) const {
+  return std::make_shared<JoinNestedLoop>(_input_left->recreate(args), _input_right->recreate(args), _mode, _column_ids,
+                                          _scan_type);
+}
+
+std::shared_ptr<const Table> JoinNestedLoop::_on_execute() {
+  PerformanceWarning("Nested Loop Join used");
+
+  _create_table_structure();
+
+  _left_column_id = _column_ids.first;
+  _right_column_id = _column_ids.second;
+
+  // Resolve both column data types and execute the main join loop
+  resolve_data_type(_left_in_table->column_type(_left_column_id), [&](auto left_type) {
+    using LeftType = typename decltype(left_type)::type;
+
+    resolve_data_type(_right_in_table->column_type(_right_column_id), [&](auto right_type) {
```

Comment:
I think resolving the type could happen at a later point. A lot of code in `_perform_join` doesn’t actually need the types. I suggest to reorder the code it bit:

```
- _perform_join (not templated)
  - loop over left chunks
    - loop over right chunks
      - resolve left type and right type
        - filter column type combinations (see other comment)
        - get comparator
        - iterables for each
          - hot loop
```

Reply:
Changed. Performance does not seem to be affected. Not sure about compile time.

Reply:
This is just a compile time optimisation.